### DISCUSSION TITLE
Discussion/Investigation on Globebrowsing Caching

### DISCUSSION THREAD

**Discussion by GPayne** [OpenSpace/OpenSpace#2125] [General]
The Utah WMS server was recently hit with a high volume of globebrowsing tile requests by an instance of OpenSpace that had been running for a few days.

The sustained level of requests was misinterpreted as a Denial-of-Service attack, in part because it happened to coincide with an actual attack on some of the other servers at SCI at the University of Utah.

The WMS server log showed that the vast majority of the requests were invalid because they used the wrong tile values (this will probably become a separate issue).

There are instances where the globebrowsing caching needs to be disabled, but we should revisit the pros & cons of this and consider enabling it as the default setting.

**Comment by alexanderbock** [comment]
Copy from Slack:
Just some brain dumping here:
The main reason not to enable it by default is that the defaults get a bit tricky;   the way we have it implemented right now is if you enable WMS caching in the config file, we add a piece of XML to every WMS server telling GDAL to cache using x GB per dataset.  The problem then becomes that we don't have control over the total amount of disk space that we are using.   Combining that with the fact that the Earth alone uses 5 layers by default (ESRI color + GIBS color, height layer, night texture, water mask) that means we get 5x GB just for a day.  Adding to this is that we treat every day of the GIBS layer as a separate layer, meaning that if you go backwards a week, you would end up with 5x + 6x GB of cache without us having any control over the maximum size that this folder will end up with.
One way to keep it more in check would be to not have the time-varying datasets be cached, but then maybe that's exactly what a user wants to happen instead

**Comment by hedbergj** [comment]
One other aspect that would help me: It's hard to know what's causing hiccups in the loading of tiles. Is it my mediocre home internet? The server where it's hosted? Or OpenSpace globebrowsing/rendering speeds? Making the following timelapse, even with Cache enabled, took way longer than I would have guessed. Is there a way I can tell where the slow down is happening?

https://user-images.githubusercontent.com/7926907/126903181-e9fe198d-9825-45bb-8db1-54a6c8d571a4.mp4

**Comment by GPayne** [comment]
This is something I've wondered about also. There are a lot of steps in-between the initial tile request and the receipt of the tile. @hedbergj , I would be willing to try rendering your timelapse at my computer at SCI which is also where our WMS server is, so I'd have the fastest possible connection. If you had a session recording prepared, you could time it from start to finish, and then I could do the same here. There still could be hardware performance differences, but it would be interesting to see how the timing compared. It could at least address the question about your home connection speed.

**Comment by hedbergj** [comment]
Thanks @GPayne  - I don't think you'd need to do the whole timelapse thing. It's just loading 7 days of VIIRS temporal (so it's not your servers either) . Each day can take a few minutes to get all the tiles in place and at the right resolution. Here's a sped up video (5 min -> 15 seconds - watch the realtime clock in the bottom) to show the loading of two days. I don't think my home internet is the problem (but I won't rule it out given the opaque nature of those services, but usually it's not an issue for most normal activities). When I made the video above in the first comment, I just manually advanced a day, waited 5-10 minutes, then took another screenshot. 

https://user-images.githubusercontent.com/7926907/127215335-a2742c55-4132-4670-a6fa-d385c9fe0d62.mp4

**Comment by hedbergj** [comment]
And, if I look at the VIIRS on here, it's way way faster, so I don't think it's my internet speed: https://worldview.earthdata.nasa.gov/?v=-124.0629695934306,37.666845602645,-117.73529938252973,42.127594910141326&l=Reference_Labels_15m,Reference_Features_15m,Coastlines_15m,VIIRS_SNPP_CorrectedReflectance_TrueColor,MODIS_Aqua_CorrectedReflectance_TrueColor(hidden),MODIS_Terra_CorrectedReflectance_TrueColor(hidden)&lg=false&t=2021-07-22-T02%3A00%3A00Z

**Comment by GPayne** [comment]
@hedbergj I recreated the setup & view from your video at work with a fast internet connection. It took 90 seconds for the full-resolution tiles to load vs. 150 seconds for yours. I agree it's not the internet speed. I'd have to add some logging to get more information about where the time is being spent.

**Comment by hedbergj** [comment]
Hi @GPayne - I just tried it again this morning in the dome with the PC and institutional grade ethernet and wouldn't you know, it only took a few seconds to get the tiles for VIIRS. Basically only a little slower than worldview link above. I could flip through the days and it was basically fine - good enough for a live show at least.

**Comment by alexanderbock** [comment]
I think there might be a lingering bug in the system that will cause issues if a tile is not returned within one frame (meaning the tile gets requested on tile X and is not there on X+1, at which point it will be requested again).  Not sure how GDAL handles that, but we are not having a protective layer on our end for that either

**Comment by GPayne** [comment]
The test I ran was on a machine with a ~700Mbps up/down (simple google speed test) ethernet connection, and I still got the 90 second load time.
